docs/proposals/registry/index-server-REST-API.md

```diff
@@ -0,0 +1,239 @@
+# Index server REST APIs
+This documentation explains how to use the index server REST APIs
+
+## Gets registry index file  
+### HTTP request
+```
+GET http://{registry host}/index
+```
+### Request parameters
+| Parameter | Description |
+| --------- | ----------- |
+| registry host | the URL/ingress that exposes registry service |
+### Request body
+The request body must be empty.
+### Request example
+```
+curl http://devfile-registry.192.168.1.1.nip.io/index
+```
+### Response example
+```
+[
+  {
+    "name": "java-maven",
+    "displayName": "Maven Java",
+    "description": "Upstream Maven and OpenJDK 11",
+    "tags": [
+      "Java",
+      "Maven"
+    ],
+    "projectType": "maven",
+    "language": "java",
+    "links": {
+      "self": "devfile-catalog/java-maven:latest"
```

Comment:
Do we have any specification for the index response, or is that part of a separate spec? I believe these self-links should match up with the example below, at least (e.g. `devfiles/nodejs`)

Reply:
Yes, these self-links are from the index that is generated from current default registry https://github.com/devfile/registry, and the example below uses the current default registry as well. Regarding the generic specification of index/index response, it is in index generator module: https://github.com/devfile/registry-support/blob/master/index/generator/schema/schema.go

Reply:
So, is it a valid self-link? Or it should be 
```suggestion
      "self": "/devfiles/java-maven"
```
?

Reply:
I will create another PR to update index specification in index generator module.

Reply:
My question (and I think what Serhii is getting at as well) is if the listed example's links are examples of what they will be in reality or just placeholders. The confusion is -- from these samples -- I would try a flow like
1. `curl http://devfile-registry.192.168.1.1.nip.io/index`
2. Get a self-link, e.g. `devfile-catalog/java-maven:latest` 
3. Try
    ```
    curl http://devfile-registry.192.168.1.1.nip.io/devfile-catalog/java-maven:latest
    ```
    to get the devfile I want.

However, `http://devfile-registry.192.168.1.1.nip.io/devfile-catalog/` will be an invalid path per the spec below:
```
GET http://{registry host}/devfiles/{stack}
```

Reply:
@amisevsk The self-link points to the stack devfile image (eg. docker image) instead of stack devfile content, so it's not related to and it's not used by the REST APIs in this PR, the index server can handle the self-link internally to access the stack devfile image.

Consumers or index server REST APIs users don't need to derive the URL to get the devfile content by using self-links, just follow docs `GET http://{registry host}/devfiles/{stack}` to update `registry host` and `stack` would be fine. For example, if user wants to get nodejs devfile content, just run `curl http://devfile-registry.192.168.1.1.nip.io/devfiles/nodejs`.

Reply:
Ah, my confusion is because my assumption upon seeing a "self" link in an index would be to assume that it's a link to the file being indexed (i.e. the devfile). If it's a link to the image, why not key the link via `"image"` or something similar? Even though it's simple, `http://{registry host}/devfiles/{stack}` is more complicated than a self link in the index (and could break if for whatever reason this spec is changed). 

What will `curl http://devfile-registry.192.168.1.1.nip.io/devfile-catalog/java-maven:latest` return?

Reply:
I am not sure why originally we name it "self" instead of "image" and I agree with you it's a little bit misleading, to update it we can change the index generator as index generator generates the "self" field, we make the changes after we confirm there is no breaking changes.

IMO, the complexity of `http://{registry host}/devfiles/{stack}` and `http://{registry host}/{self}` should be the same because the difference between the two ways is to get {stack} vs to get {self}, and both {stack} and {self} can be get from index under specific stack with one query, so the complexity would be the same. However, I think `http://{registry host}/devfiles/{stack}` is more RESTful, which clearly locates the resource by using nouns with different levels.

`curl http://devfile-registry.192.168.1.1.nip.io/devfile-catalog/java-maven:latest` is invalid, because according to this index server REST APIs docs, currently we only serve two REST APIs below:
- To get index content: `GET http://{registry host}/index`
- To get devfile content: `GET http://{registry host}/devfiles/{stack}`

Reply:
> IMO, the complexity of http://{registry host}/devfiles/{stack} and http://{registry host}/{self} should be the same because the difference between the two ways is to get {stack} vs to get {self},

The difference here is that `http://{registry host}/devfiles/{stack}` is a convention we define in our spec, whereas `http://{registry host}/{self}` (where `{self}` is `devfiles/{stack}`) is provided to the consumer at the time of getting the index. If we go with the former, all users of the registry will need to be updated if there's e.g. a different structure ( `http://{registry host}/devfiles/v3/{stack}`). With the latter, the registry can put devfiles where it likes and consumers don't need to worry about it.

Reply:
Just chiming in here, initially we went with `self` as it meant a minimal number of changes to index.json, and was quick to prototype. I agree that we should change it, though I would say `stack` would be my preference over `image`.

Reply:
> The difference here is that http://{registry host}/devfiles/{stack} is a convention we define in our spec, whereas http://{registry host}/{self} (where {self} is devfiles/{stack}) is provided to the consumer at the time of getting the index. If we go with the former, all users of the registry will need to be updated if there's e.g. a different structure ( http://{registry host}/devfiles/v3/{stack}). With the latter, the registry can put devfiles where it likes and consumers don't need to worry about it.

I get your point, I agree that if the API versions get involve into this, to use the `self` can avoid users to specify the version explicitly, but I am wondering if the API supports multiple versions and when users want to choose which version to use, then we might still need to provide the version option somewhere (eg. selfv2, selfv3) and users still need to update that as well.

To make the change, I think the URL of the REST API design is still `http://{registry host}/devfiles/{stack}`, but if we do need a "self" as the shortcut for users we can the following things:
- Change the **current** `self` field to `image` or `stack`
- Add the `self` field to point to `devfiles/{stack}`

Which means the main URL definition is `http://{registry host}/devfiles/{stack}`, users can use either use `http://{registry host}/devfiles/{stack}` or `http://{registry host}/{self}` to get the content of the devfile, so that we can both provide a clear URL definition to users with RESTful design and let users get the devfile content with `self` placeholder without worrying about the structure, we can use separate issue and PR to implement that.

Reply:
@amisevsk Issue has been open: https://github.com/devfile/api/issues/297